server/src/main/java/org/opensearch/common/util/LongRHHash.java

```diff
@@ -0,0 +1,242 @@
+     */
+    private final long[] hints;
+
+    /**
+     * Bit-mask used to calculate the slot in the hints' table.
+     */
+    private final int hintMask;
+
+    public LongRHHash(BigArrays bigArrays) {
+        this(DEFAULT_CAPACITY, DEFAULT_LOAD_FACTOR, DEFAULT_HINTS, bigArrays);
+    }
+
+    public LongRHHash(long capacity, float loadFactor, int numHints, BigArrays bigArrays) {
+        capacity = (long) (capacity / loadFactor);
+        capacity = Math.max(1, Long.highestOneBit(capacity - 1) << 1); // next power of two
+
+        this.loadFactor = loadFactor;
+        this.bigArrays = bigArrays;
+        this.capacity = capacity;
+        this.mask = capacity - 1;
+        this.size = 0;
+        this.grow = (long) (capacity * loadFactor);
+
+        this.ords = bigArrays.newLongArray(capacity, false);
+        this.ords.fill(0, capacity, -1);
+        this.keys = bigArrays.newLongArray(capacity, false);
+
+        numHints = Math.max(1, Integer.highestOneBit(numHints - 1) << 1);
+        this.hintMask = numHints - 1;
+        this.hints = new long[numHints];
+        Arrays.fill(this.hints, -1);
+    }
+
+    /**
+     * Adds the key to the hash table and returns its ordinal (i.e. the index) starting from zero.
+     * If the key exists already, it returns (-1-ordinal).
+     */
+    public long add(final long key) {
+        final long ordinal = find(key);
+        if (ordinal != -1) {
+            return -(1 + ordinal);
+        }
+
+        if (size >= grow) {
+            grow();
+        }
+        return insert(key, size);
+    }
+
+    /**
+     * Associates the given key with the given ordinal.
+     *
+     * This is inspired by Robin Hood hashing, but not implemented wholly.
+     * Keys are moved around dynamically in order to reduce the longest probe sequence length (PSL),
+     * which makes lookups faster as keys are likely to be found in the same CPU cache line.
+     */
+    private long insert(final long key, final long ordinal) {
+        long idx = slot(key);   // The ideal "home" location for the given key in the ordinals' table.
+        long ord = ordinal;     // The ordinal yet to find an empty slot in the ordinals' table (candidate).
+        long psl = 0;           // The displacement of "ord" from its home location.
+
+        long curOrd;            // The current ordinal stored at "idx".
+        long curPsl;            // The current ordinal's PSL.
+
+        do {
+            if ((curOrd = ords.get(idx)) == -1) {
+                // Empty slot; insert the candidate ordinal here.
+                ords.set(idx, ord);
+                keys = bigArrays.grow(keys, size + 1);
+                keys.set(ordinal, key);
+                return size++;
+            } else if ((curPsl = psl(keys.get(curOrd), idx)) < psl) {
+                // Current key is "richer" than the candidate key at this slot; swap and find a new home for it.
+                ord = ords.set(idx, ord);
```

Comment:
This sets the current ordinal in the ideal home, but doesn't move the existing one (in the ideal home) to a new one, where is that handled?

Reply:
`set(idx, ord)` sets the new value and returns the old value at that index. So our do-while loop continues to find an empty or a better slot for `ord`.

Reply:
Got it! thanks!